lib/exercises.rb

```diff
@@ -1,19 +1,64 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n * m) => where n is the number of strings & m is the length of each word
+# Space Complexity: O(n) => because of hash
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  words = {} 
+  
+  until strings.empty? 
+    i = 0 
+    word_to_compare = strings[0] # "eat" => "tan" => "bat"
+    words[word_to_compare] = []
+    # ["eat", "tea", "ate"]
+    # ["tan", "nat"]
+    # ["bat"]
+
+    while i < strings.length 
+      if permutation(word_to_compare, strings[i])
+        words[word_to_compare].push(strings[i])
+        strings.delete_at(i) 
+        i -= 1
+      end
+      i += 1 # 0 => 0
+    end
+  end
+
+  return words.values
+end
+
+def permutation(str1, str2)
+  return true if str1.empty? && str2.empty?
+  return false if str1.nil? || str2.nil?
+  return false if str1.empty? || str2.empty?
+
+  hash = {}
+
+  str1.each_char do |char|
+    hash[char] ? hash[char] += 1 : hash[char] = 1
+  end
+
+  str2.each_char do |char|
+    hash[char] ? hash[char] -= 1 : false
+  end
+
+  return hash.values.all? { |value| value == 0 }
 end
 
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n) => where n is the size of list
+# Space Complexity: O(n) => because I created a hash and an array
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  return [] if k == 0
+
+  frequency = {}
+  list.each do |element|
+    frequency[element] ? frequency[element] += 1 : frequency[element] = 1
+  end
+
+  return frequency.keys[0...k]
```

Comment:
I think this is only getting the 1st `k` keys.  We want the `k` most frequent keys.  You would need to sort the most frequent keys.

Reply:
Oops thanks for catching this! :D